x/stake/keeper/slash.go

```diff
@@ -4,56 +4,212 @@ import (
+		k.SetUnbondingDelegation(ctx, unbondingDelegation)
+		pool := k.GetPool(ctx)
+		// Burn loose tokens
+		// Ref https://github.com/cosmos/cosmos-sdk/pull/1278#discussion_r198657760
+		pool.LooseTokens -= slashAmount.Int64()
+		k.SetPool(ctx, pool)
+	}
+
+	return
+}
+
+// slash a redelegation and update the pool
+func (k Keeper) slashRedelegation(ctx sdk.Context, validator types.Validator, redelegation types.Redelegation, infractionHeight int64, fraction sdk.Rat) (slashAmount sdk.Int) {
+	now := ctx.BlockHeader().Time
+
+	// If redelegation started before this height, stake didn't contribute to infraction
+	if redelegation.CreationHeight < infractionHeight {
+		return sdk.ZeroInt()
+	}
+
+	if redelegation.MinTime < now {
+		// Redelegation no longer eligible for slashing, skip it
+		// TODO Delete it automatically?
+		return sdk.ZeroInt()
+	}
+
+	// Calculate slash amount proportional to stake contributing to infraction
+	slashAmount = sdk.NewRatFromInt(redelegation.InitialBalance.Amount, sdk.OneInt()).Mul(fraction).EvaluateInt()
+
+	// Don't slash more tokens than held
+	redelegationSlashAmount := slashAmount
+	if redelegationSlashAmount.GT(redelegation.Balance.Amount) {
+		redelegationSlashAmount = redelegation.Balance.Amount
+	}
+
+	// Update redelegation if necessary
+	if !redelegationSlashAmount.IsZero() {
+		redelegation.Balance.Amount = redelegation.Balance.Amount.Sub(redelegationSlashAmount)
+		k.SetRedelegation(ctx, redelegation)
+	}
+
+	// Unbond from target validator
+	sharesToUnbond := fraction.Mul(redelegation.SharesDst)
+	if !sharesToUnbond.IsZero() {
+		delegation, found := k.GetDelegation(ctx, redelegation.DelegatorAddr, redelegation.ValidatorDstAddr)
+		if !found {
+			// If deleted, delegation has zero shares, and we can't unbond any more
+			return slashAmount
+		}
+		if sharesToUnbond.GT(delegation.Shares) {
+			sharesToUnbond = delegation.Shares
+		}
+		tokensToBurn, err := k.unbond(ctx, redelegation.DelegatorAddr, redelegation.ValidatorDstAddr, sharesToUnbond)
+		if err != nil {
+			panic(fmt.Errorf("error unbonding delegator: %v", err))
+		}
+		// Burn loose tokens
+		pool := k.GetPool(ctx)
+		pool.LooseTokens -= tokensToBurn
+		k.SetPool(ctx, pool)
+	}
+
+	return slashAmount
+}
+
+// Slash a validator for an infraction committed at a known height
+// Find the contributing stake at that height and burn the specified fraction
+// of it, updating unbonding delegation & redelegations appropriately
+//
+// CONTRACT:
+//    Validator exists and can be looked up by public key
+// CONTRACT:
+//    Infraction committed equal to or less than an unbonding period in the past,
+//    so all unbonding delegations and redelegations from that height are stored
+// CONTRACT:
+//    Infraction committed at the current height or at a past height,
+//    not at a height in the future
+func (k Keeper) Slash(ctx sdk.Context, pubkey crypto.PubKey, infractionHeight int64, power int64, fraction sdk.Rat) {
```

Comment:
Could this function be placed above the previous two? (Sorry for the extremely minor nitpick here. I think it would be useful structurally to see the contract first though)

Also can fraction be renamed to `slashFactor`

Reply:
Moved, now `slashFactor`.